### PR TITLE
feat: add equality dunder method/ improve copy method

### LINKED ISSUE #785: Performance: Faster network.copy(), perhaps using copy.deepcopy
At present, copying a network with `pypsa.Network().copy()` takes a surprisingly long time. For some workflows this actually seems to be a performance bottleneck.

I noticed that using the builtin `copy.deepcopy` seems to do the same job much faster:

```python
import pypsa
import copy

n = pypsa.examples.ac_dc_meshed()
n2 = n.copy()  # takes about 158ms
n3 = copy.deepcopy(n)  # takes about 5 ms
```

Would this be a suitable alternative implementation? Are there any issues with using `copy.deepcopy`?

I can look into making a PR if that would be of interest.

### PR DESCRIPTION
**Pull request by lkstrp**
Closes #785 

## Changes proposed in this Pull Request
- Adds `__eq__` dunder method to Network class to compare two networks equality operator (``==``). Very helpful for future testing. We could also add a `n.equals` method similar too `pandas`, if needed
- Uses `copy.deepcopy()` as default behaviour in `n.copy()` since it is faster for most networks (1x - 10x), depending on size. But still allows for filtering, when needed
- Fixes bug in old `n.copy()` behaviour, where some network attrs have not been copied
- Fixes bug in `n.set_snapshots()`, where snapshots for non default attributes are not filtered

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

File: pypsa/components.py [hunk 1]
```diff
@@ -375,6 +379,54 @@ def __add__(self, other):
...
+                    return False
+            # Iterators
+            elif isinstance(a, (dict, Dict)):
+                for k, v in a.items():
+                    if not equals(v, b[k]):
+                        return False
+            elif isinstance(a, (list, tuple)):
+                for i, v in enumerate(a):
+                    if not equals(v, b[i]):
+                        return False
```
Comment (FabianHofmann) [inline pypsa/components.py:402]: dicts, list and tuples you can directly compare with `==` without having to loop over single keys, elements

File: pypsa/components.py [hunk 2]
```diff
@@ -375,6 +379,54 @@ def __add__(self, other):
...
+            else:
+                if a != b:
+                    return False
+
+            return True
+
+        if isinstance(other, self.__class__):
+            for key, value in self.__dict__.items():
+                if not equals(value, other.__dict__[key]):
+                    return False
```
Comment (FabianHofmann) [inline pypsa/components.py:420]: same here

File: pypsa/components.py [hunk 3]
```diff
@@ -576,11 +633,15 @@ def set_snapshots(
...
             attrs = self.components[component]["attrs"]
 
-            for k, default in attrs.default[attrs.varying].items():
+            for k in pnl.keys():
                 if pnl[k].empty:  # avoid expensive reindex operation
                     pnl[k].index = self._snapshots
+                elif k in attrs.default[attrs.varying]:
+                    pnl[k] = pnl[k].reindex(
+                        self._snapshots, fill_value=attrs.default[attrs.varying][k]
+                    )
```
Comment (FabianHofmann) [inline pypsa/components.py:642]: this change is purely stylistic or?

File: pypsa/components.py [hunk 4]
```diff
@@ -996,34 +1056,80 @@ def _retrieve_overridden_components(self):
...
+                f"instead. This will raise an error in a future version.",
+                UserWarning,
+                stacklevel=2,
+            )
+
+        # Use copy.deepcopy if no arguments are passed
+        args = [snapshots, investment_periods, ignore_standard_types, with_time]
+        if all(arg is None or arg is False for arg in args):
+            logger.info("Creating a deep copy of the network.")
+            return copy.deepcopy(self)
```
Comment (FabianHofmann) [inline pypsa/components.py:1130]: this is great!

File: pypsa/components.py [hunk 5]
```diff
@@ -1035,40 +1141,60 @@ def copy(
             override_component_attrs=override_component_attrs,
         )
 
+        # Copy components
         other_comps = sorted(self.all_components - {"Bus", "Carrier"})
         for component in self.iterate_components(["Bus", "Carrier"] + other_comps):
```
Comment (FabianHofmann) [inline pypsa/components.py:1146]: nothing, you have changed but what looks weird. why not saying 
```python
        for component in self.iterate_components(self.all_components):
```

**Review by FabianHofmann — COMMENTED** [review]
looks good, just had a short look, here some comments

File: pypsa/components.py (hunk 3 above)
Comment (lkstrp) [reply pypsa/components.py:642]: No, previously new snapshots have not been applied to custom attributes, so non-default attributes

File: pypsa/components.py (hunk 1 above)
Comment (FabianHofmann) [reply pypsa/components.py:402]: gotcha

File: pypsa/components.py (hunk 3 above)
Comment (FabianHofmann) [reply pypsa/components.py:642]: :ok_hand:

**Pull request merged**